CommunityToolkit.Authentication.Uwp/WindowsProvider.cs

```diff
@@ -562,24 +562,37 @@ private async Task<List<WebAccountProvider>> GetWebAccountProvidersAsync()
             if (_webAccountProviderConfig.WebAccountProviderType == WebAccountProviderType.Any ||
                 _webAccountProviderConfig.WebAccountProviderType == WebAccountProviderType.Msa)
             {
-                providers.Add(await WebAuthenticationCoreManager.FindAccountProviderAsync(MicrosoftProviderId, MicrosoftAccountAuthority));
+                await FindAndAddProviderAsync(MicrosoftProviderId, MicrosoftAccountAuthority);
             }
 
             // AAD
             if (_webAccountProviderConfig.WebAccountProviderType == WebAccountProviderType.Any ||
                 _webAccountProviderConfig.WebAccountProviderType == WebAccountProviderType.Aad)
             {
-                providers.Add(await WebAuthenticationCoreManager.FindAccountProviderAsync(MicrosoftProviderId, AadAuthority));
+                await FindAndAddProviderAsync(MicrosoftProviderId, AadAuthority);
             }
 
             // Local
             if (_webAccountProviderConfig.WebAccountProviderType == WebAccountProviderType.Any ||
                 _webAccountProviderConfig.WebAccountProviderType == WebAccountProviderType.Local)
             {
-                providers.Add(await WebAuthenticationCoreManager.FindAccountProviderAsync(LocalProviderId));
+                await FindAndAddProviderAsync(LocalProviderId);
             }
 
             return providers;
+
+            async Task FindAndAddProviderAsync(
+                string webAccountProviderId,
+                string authority = default)
+            {
+                var provider = string.IsNullOrEmpty(authority)
+                    ? await WebAuthenticationCoreManager.FindAccountProviderAsync(webAccountProviderId)
+                    : await WebAuthenticationCoreManager.FindAccountProviderAsync(webAccountProviderId, authority);
+                if (provider != null)
+                {
+                    providers.Add(provider);
+                }
+            }
```

Comment:
Would it be nicer to just add a private extension method for `AddIfNotNull` and use that in the original code? May be easier to read and understand then burring the `WebAuthenticationManager` and doing the ternary operator based on parameters?

@Sergio0694 would we ever want to add LINQ helpers like that to the base .NCT?

Reply:
Not sure how that extension would avoid the ternary operator here, given it's calling a different constructor based on that. I don't personally have an issue with this snippet though, it seems fine to me. Only nit, I'd use `is not null` 😄

> "would we ever want to add LINQ helpers like that to the base .NCT?"

I'm personally against having such small extensions that basically only forward to the built-in API or do a single small thing. I feel like they actually make the code harder to read, as now you'll end up in a situation where half the devs would just use normal C#, and half would instead use these random extensions. We've rejected similar proposals in the past as well, where eg. someone proposed a LINQ `None` extension that literally just did `!Any`.

Imho, just having a small `if` like this is perfectly fine 🙂

Reply:
> Not sure how that extension would avoid the ternary operator here, given it's calling a different constructor based on that. I don't personally have an issue with this snippet though, it seems fine to me. Only nit, I'd use `is not null` 😄

As it'd be encapsulated in the original call like so:

```cs
providers.AddIfNotNull(await WebAuthenticationCoreManager.FindAccountProviderAsync(MicrosoftProviderId, MicrosoftAccountAuthority));
```

And then the parameter is only added to the collection if it's not `null` in the extension, so you don't have to do the parameter passing to the inner function, check the arguments with the ternary, and then break out of the function scope to add to the locally defined list.

> I'm personally against having such small extensions that basically only forward to the built-in API or do a single small thing. I feel like they actually make the code harder to read, as now you'll end up in a situation where half the devs would just use normal C#, and half would instead use these random extensions. We've rejected similar proposals in the past as well, where eg. someone proposed a LINQ `None` extension that literally just did `!Any`.

Yeah, if it's an inversion, then I agree it's not as useful. Here there's a bit of control logic flow which I feel could make it easier to flow/guard against these cases which could be wrapped up nicely. Will have to see how the new extension and `role` features hopefully coming in the near future will change things if we have more `static` based helpers we want to add anywhere else as well.

For now, I guess we can leave as-is.

Reply:
Mmmh not sure, I'll get back to you on this.


Re:

```csharp
providers.AddIfNotNull(await WebAuthenticationCoreManager.FindAccountProviderAsync(MicrosoftProviderId, MicrosoftAccountAuthority));
```

I'm... Not sure I'm following here, would this not be different than the code in the PR? The code with the ternary operator is using two different APIs to resolve an object, with different arguments (ie. also passing the authority or not depending on whether it's null or not). And _then_ there's a branch on whether this resulting object is null or not, to add it to the collection. Isn't the snippet you used functionally different, since it always uses the API taking both the provider and the authority? 🤔